### ISSUE TITLE
Self consistency of hypre_CSRMatrix

### ISSUE THREAD

**Issue by v-dobrev** [hypre-space/hypre#143]
In the `hypre_CSRMatrix` struct:
https://github.com/hypre-space/hypre/blob/3b22847deb3431416604b3b3a3645cd987c2eda9/src/seq_mv/csr_matrix.h#L23-L46

isn't it more appropriate to switch the type of `num_cols` from `HYPRE_Int` to `HYPRE_BigInt`? I'm thinking of the case when hypre is built with the "mixed-int" option, and a `hypre_CSRMatrix` object uses the `big_j` array -- in that case the column indices are 64-bit but the number of columns, `num_cols`, is just 32-bit.

**Comment by ulrikeyang** [comment]
I think if we really get a case for which we would need 64bit integers for the number of columns which is really a local number connected to the actual number of coefficients we need to store we would be in big trouble. It should be fine for this to be 32 bits. I don’t see a case where this number should ever get this big. Maybe some day we will have machines that are that big and could handle that.


---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>

On June 25, 2020 at 12:53:12 AM PDT, Veselin Dobrev <notifications@github.com> wrote:

In the hypre_CSRMatrix struct:
https://github.com/hypre-space/hypre/blob/3b22847deb3431416604b3b3a3645cd987c2eda9/src/seq_mv/csr_matrix.h#L23-L46

isn't it more appropriate to switch the type of num_cols from HYPRE_Int to HYPRE_BigInt? I'm thinking of the case when hypre is built with the "mixed-int" option, and a hypre_CSRMatrix object uses the big_j array -- in that case the column indices are 64-bit but the number of columns, num_cols, is just 32-bit.

**Comment by v-dobrev** [comment]
I think one use case (I thought that was the reason to have `big_j` in the first place) is in the function `hypre_MergeDiagAndOffd` where after the merge the column indices are global indices, so they need to be 64-bit (in `HYPRE_BIGINT` and `HYPRE_MIXEDINT` modes).

**Comment by ulrikeyang** [comment]
Yes, mergediagandoffd is a problem which we need to revisit.


---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>

On June 27, 2020 at 7:56:39 PM PDT, Veselin Dobrev <notifications@github.com> wrote:

I think one use case (I thought that was the reason to have big_j in the first place) is in the function hypre_MergeDiagAndOffd where after the merge the column indices are global indices, so they need to be 64-bit (in HYPRE_BIGINT and HYPRE_MIXEDINT modes).

**Comment by victorapm** [comment]
I was looking at cleaning hypre's compilation with the `-Wconversion` flag and found some examples where storing the number of columns as a `HYPRE_BigInt` instead of just `HYPRE_Int` seems to be required. See:

https://github.com/hypre-space/hypre/blob/a09d776fd2f8453bb517120cc2ae3730d18a93ee/src/parcsr_mv/par_csr_matop.c#L5090
https://github.com/hypre-space/hypre/blob/a09d776fd2f8453bb517120cc2ae3730d18a93ee/src/parcsr_mv/par_csr_aat.c#L940
https://github.com/hypre-space/hypre/blob/a09d776fd2f8453bb517120cc2ae3730d18a93ee/src/parcsr_ls/ams.c#L4254-L4256
https://github.com/hypre-space/hypre/blob/a09d776fd2f8453bb517120cc2ae3730d18a93ee/src/parcsr_ls/ams.c#L4430-L4432

These code snippets share something in common: they are all creating a `hypre_CSRMatrix` and passing the number of columns as a `HYPRE_BigInt`, however the function itself expects a `HYPRE_Int`, therefore there's a implicit conversion from `HYPRE_BigInt` to `HYPRE_Int` which won't be correct when the number exceeds INT32_MAX.

How should we handle this issue?

**Comment by liruipeng** [comment]
I would say this is just "okay" as is, since we don't use the number of columns in any of the cases (which is a bigint but assigned to a int, etc...). To make it exactly correct, we will need to have a bigint numcol and probably enforcing mutually exclusive J and bigJ. We talked about it in #623. Happy to revisit this. Thanks!

**Comment by victorapm** [comment]
OK! So if we are fine with possible loss of data here, I would propose to explicitly pass the number of columns as a `HYPRE_Int`, i.e., instead of doing:
```c
 hypre_CSRMatrix *local = hypre_CSRMatrixCreate (num_edges, 
                                                 num_global_vert, 
                                                 2 * num_edges); 
```
we would do
```c

 hypre_CSRMatrix *local = hypre_CSRMatrixCreate (num_edges, 
                                                 (HYPRE_Int) num_global_vert, 
                                                 2 * num_edges); 
```

**Comment by liruipeng** [comment]
Yeah, that is actually what the code implicitly does. Again, I am not totally happy with it though, so willing to discuss the changes.
